### DISCUSSION TITLE
Using request vs fetch

### DISCUSSION THREAD

**Discussion by simhnna** [nodejs/undici#3253] [Polls]
Assuming the code will always (TM) be executed with node, never in browser or another js runtime. Should I use fetch or request?
I'm tending towards request, but fetch being bundled and the compatibility is tempting...

# fetch
## pro
* available globally
* all libs can use the same implementation
* bundled in node
* browser fetch compatibility
## contra
* error handling (DomException, errors being wrapped in TypeError, ...)

# request
## pro
* better error handling (the errors are clearer)
* more control over timeouts
## contra
* need a dependency on undici (additional dependency, don't know if it will become incompatible to bundled node)
* If for some reason I decide to use code somewhere else I'd need to refactor request -> fetch
* doesn't handle compression

**Comment by Uzlopak** [comment]
fetch handles compression. for request you have to implement it yourself.

Also there are still some improvements for fetch in the pipeline. like caching. Stuff you would need to implement yourself in request.

**Reply by simhnna** [reply]
oh. I wasn't aware of compression. I guess that's close to being a deal breaker for me..

**Comment by Yogu** [comment]
Is the error handling something that can be improved in the `fetch` implementation? I think specs usually only dictate the `name` of an error, not the message.

**Reply by simhnna** [reply]
after playing around a little more it looks like fetch is always wraps an error (usually an undici error) in a `TypeError: fetch failed`

There's one exception. If a request is aborted using an AbortSignal both fetch and request throw ` DOMException [TimeoutError]: The operation was aborted due to timeout`

**Comment by metcoder95** [comment]
The overall recommendation we give is to use `undici.request` when possible as it is not only the most performant but also fine grained control over your request and gives up-to-date advantage of `undici`.

They way `undici` is bundled is meant to empower `fetch` and do not collide with userland `undici`. 

If you are really seeking for interoperability and dependency-less option, `fetch` is the way to go; but for the rest I always recommend `undici.request` or `undici` overall

**Reply by ronag** [reply]
We should probably add a decompression interceptor to handle that case as well.
